vm-virtio/src/vhost_user/net.rs

```diff
@@ -90,51 +92,56 @@ impl Net {
+            | 1 << VIRTIO_F_EVENT_IDX
+            | VhostUserVirtioFeatures::PROTOCOL_FEATURES.bits();
 
         // Get features from backend, do negotiation to get a feature collection which
         // both VMM and backend support.
         let backend_features = vhost_user_net.get_features().unwrap();
         avail_features &= backend_features;
         // To set features back is decided by the vhost crate mechanism, since the
         // later vhost call requires backend_features filled in master,
-        // which is setup by the call here. Will check if the corresponding logic
-        // in vhost is sensible in the future.
+        // which is setup by the call here.
         vhost_user_net
             .set_features(backend_features)
             .map_err(Error::VhostUserSetFeatures)?;
 
-        // Enable the following features here which are not need to be negotiated with
-        // vhost-user backends.
-        avail_features |= 1 << virtio_net::VIRTIO_NET_F_MAC
-            | 1 << virtio_net::VIRTIO_NET_F_CTRL_VQ
-            | 1 << virtio_net::VIRTIO_NET_F_MQ;
-
-        // Fill config space which has mac_addr, status and MQ info currently.
         let mut config_space = mac_addr.get_bytes().to_vec();
+        avail_features |= 1 << virtio_net::VIRTIO_NET_F_MAC;
         if num_queue_pairs as u16 >= VIRTIO_NET_CTRL_MQ_VQ_PAIRS_MIN
             && num_queue_pairs as u16 <= VIRTIO_NET_CTRL_MQ_VQ_PAIRS_MAX
         {
             config_space.resize(CONFIG_SPACE_VUNET, 0);
             let max_queue_pairs = (num_queue_pairs as u16).to_le_bytes();
             config_space[CONFIG_SPACE_MAC + CONFIG_SPACE_STATUS..CONFIG_SPACE_VUNET]
                 .copy_from_slice(&max_queue_pairs);
+            avail_features |=
+                1 << virtio_net::VIRTIO_NET_F_CTRL_VQ | 1 << virtio_net::VIRTIO_NET_F_MQ;
         }
 
+        let protocol_features;
         let mut acked_features = 0;
-        let max_queue_number = if backend_features & (1 << VHOST_USER_F_PROTOCOL_FEATURES) != 0 {
-            acked_features |= 1 << VHOST_USER_F_PROTOCOL_FEATURES;
-            let protocol_features = vhost_user_net.get_protocol_features().unwrap();
-            vhost_user_net
-                .set_protocol_features(protocol_features)
-                .map_err(Error::VhostUserSetProtocolFeatures)?;
-            match vhost_user_net.get_queue_num() {
-                Ok(qn) => qn,
-                Err(_) => num_queues as u64,
-            }
+        if backend_features & VhostUserVirtioFeatures::PROTOCOL_FEATURES.bits() != 0 {
+            acked_features |= VhostUserVirtioFeatures::PROTOCOL_FEATURES.bits();
+            protocol_features = vhost_user_net
+                .get_protocol_features()
+                .map_err(Error::VhostUserGetProtocolFeatures)?;
         } else {
-            num_queues as u64
-        };
+            return Err(Error::VhostUserProtocolNotSupport);
+        }
+
+        let max_queue_number =
```

Comment:
Can you please squash this commit directly into the one introducing this code.

Reply:
Ditto!